File: Binary_Search_Tree/530.Minimum_Absolute_Difference_in_BST/Python/530.Minimum_Absolute_Difference_in_BST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def getMinimumDifference(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        #======================================#
        # Recursion-based DFS traversal method #
        #======================================#

        ############
        #Initialize
        ##### Record array #####
        record_arry = []

        ##### Result minimun difference #####
        res_min_diff = ((10 ** 9) + 1)


        ####################################
        #Recursion-based DFS loop traversal
        self.dfsTraversal(root, record_arry) #Recursion function call (i.e. whole root tree-node)

        for record_idx in range((len(record_arry) - 1)):
            res_min_diff = min(res_min_diff, abs(record_arry[record_idx] - record_arry[(record_idx + 1)])) #Keep updating/overwriting

        return res_min_diff


    def dfsTraversal(self, root, record_arry):
        """
        :type root: Optional[TreeNode]
        :type record_arry: List[int]
        :rtype: None, void
        """
        #========================================#
        # Recursion-based DFS traversal function #
        #========================================#
        if (not root): #Issue/Boundary-case handling
            return

        ####################
        #Whole process/flow
        self.dfsTraversal((root.left), record_arry) #Recursion function call (i.e. root left-half node)

        record_arry.append(root.val) #Update/Record

        self.dfsTraversal((root.right), record_arry) #Recursion function call (i.e. root right-half node)

File: Binary_Search_Tree/530.Minimum_Absolute_Difference_in_BST/C++/530.Minimum_Absolute_Difference_in_BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int getMinimumDifference(TreeNode * root) 
    {
        //=====================================//
        // Input type:                         //
        //   - TreeNode * (root tree-node)     //
        // Return type:                        //
        //   - int (result minimun difference) //
        //=====================================//

        /*Initialize*/
        ///// Record vector /////
        vector<int> record_vect;

        ///// Result minimun difference /////
        int res_min_diff = (1e9 + 1);


        /*Recursion-based DFS loop traversal*/
        dfsTraversal(root, record_vect); //Recursion function call

        for (int record_idx = 0; (record_idx < (record_vect.size() - 1)); (record_idx++)) //Whole
        {
            res_min_diff = min(res_min_diff, abs(record_vect[record_idx] - record_vect[(record_idx + 1)])); //Keep updating, overwriting

        } //Whole

        return res_min_diff;
    }


    void dfsTraversal(TreeNode * root, vector<int> & record_vect)
    {
        //======================================//
        // Input type:                          //
        //   - TreeNode * (root tree-node)      //
        //   - vector<int> ref. (record vector) //
        // Return type:                         //
        //   - void (no return)                 //
        //======================================//
        if (! root) { return; } //Issue, Boundary-case handling

        /*Whole process, flow*/
        dfsTraversal((root->left), record_vect); //Recursion function call (i.e. root left-half node)

        record_vect.emplace_back((root->val)); //Update, Record

        dfsTraversal((root->right), record_vect); //Recursion function call (i.e. root right-half node)
    }
};
